File: Data Structures & Algorithms/three-integer-sum/submission-0.cs
public class Solution {
    public List<List<int>> ThreeSum(int[] nums) {
        Array.Sort(nums) ;
        var res = new List<List<int>>();
        for(int i =0 ; i< nums.Length ; i++){
            if(nums[i] > 0 ) break;
            if(i > 0 && nums[i] == nums[i-1]) continue ;
            int l = i + 1 ;
            int r = nums.Length -1 ;
            while(l < r){
                int ret = nums[i] + nums[l] + nums[r];
                if(ret == 0){
                    res.Add(new List<int>(){nums[i], nums[l], nums[r]});
                    r -- ;
                    l ++;
                    while(l < r && nums[l] == nums[l -1]) l++;
                }
                else if(ret > 0){
                    r-- ;
                }
                else{
                    l++;
                }
            }
        }
        return res ;
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-0.java
class Solution {

    List<List<Integer>> res = new ArrayList<>();
    public List<List<Integer>> subsetsWithDup(int[] nums) {

        Arrays.sort(nums);
        process(nums,0, new ArrayList<Integer>());
        return res ;
    }

    void process(int[] nums, int index, List<Integer> path){
  
        res.add(new ArrayList<>(path));
         

        for(int i = index ; i< nums.length ; i++){

            if(i > index && nums[i] == nums[i-1]) continue; 
            path.add(nums[i]);
            process(nums, i + 1, path);
            path.remove(path.size() -1);
        }
    }
}
